core/src/main/scala/hedgehog/package.scala

```diff
@@ -31,18 +31,86 @@ package object hedgehog extends ApplicativeSyntax {
   def propertyT: PropertyTOps =
     new PropertyTOps {}
 
+  /**
+    * Compare two arguments with the comparison function and return Result.success
+    * if the function return true. Otherwise, it returns Result.failure with
+    * the Log containing the argument values.
+    *
+    * @example
+    * {{{
+    *   val a1 = "abc"
+    *   val a2 = "abc"
+    *   diff(a1, a2)(_ == _)
+    *   // Result.success
+    *
+    *   diff(123, 456)(_ != _).log("It must be different.")
+    *   // Result.success
+    *
+    *   val x = 'q'
+    *   val y = 80
+    *   diff(x, y)((x, y) => y < 87 && x <= 'r')
+    *   // Result.success
+    * }}}
+    * @example
+    * {{{
+    *   val a1 = "abc"
+    *   val a2 = "xyz"
+    *   diff(a1, a2)(_ == _)
+    *   // Result.failure
+    *   > === Failed ===
+    *   > --- lhs ---
+    *   > abc
+    *   > --- rhs ---
+    *   > xyz
+    *
+    *   diff(123, 123)(_ != _).log("It must be different.")
+    *   // Result.failure
+    *   > === Failed ===
+    *   > --- lhs ---
+    *   > 123
+    *   > --- rhs ---
+    *   > 123
+    *   > It must be different.
+    *
+    *   val x = 'z'
+    *   val y = 100
+    *   diff(x, y)((x, y) => y < 87 && x <= 'r')
+    *   // Result.failure
+    *   > === Failed ===
+    *   > --- lhs ---
+    *   > z
+    *   > --- rhs ---
+    *   > 100
+    * }}}
+    *
+    * @param a1 The left-hand side argument
+    * @param a2 The right-hand side argument
+    * @param f The comparison function to compare the given two arguments
+    * @tparam A the type of the arguments
+    * @return Result.success. if the f(a1, a2) returns true.
```

Comment:
Sorry to be a pain, but I'm not a fan of this over-the-top doc. The type of the function is pretty much all I need to know how this works. I like the actual description and examples, I'm just think the param/return doc is not needed (as much) when you have good parametric types.

Reply:
OK, I will remove it. I was trying to make it similar to Haskell one which has that kind of comment. 🙂